Windows dependencies to build

blackshadev opened the pull request:
# What has been done

- Add building libusb and hidapi to the windows github actions workflow. 

# How to test

- Green build pipeline
- Check the "configure" step
- See all transports  enabled like below, except BLE, which isn't implemented in libdivecomputer.
```
  Transports:

    Serial    : yes
    USB       : yes
    USBHID    : yes
    IrDA      : yes
    Bluetooth : yes
    BLE       : no
```

jefdriesen reviewed (changes requested):
Nice work! This was actually something that was already on my todo list, and I just didn't have time for it yet.

My main comments are about using pkg-config when cross-compiling. There are also a couple of whitespace issue (trailing spaces).

Do you mind squashing all your changes into a single commit, to keep the history a bit tidy? I can also take care of that during merging if you prefer.

blackshadev commented:
> Nice work! This was actually something that was already on my todo list, and I just didn't have time for it yet.
> 
> My main comments are about using pkg-config when cross-compiling. There are also a couple of whitespace issue (trailing spaces).
> 
> Do you mind squashing all your changes into a single commit, to keep the history a bit tidy? I can also take care of that during merging if you prefer.

Thanks! I will process your feedback later this week and squash everything when I am done. This was quite a wild ride getting everything working. Thanks for the quick review and in-depth suggestions.

blackshadev commented:
> Nice work! This was actually something that was already on my todo list, and I just didn't have time for it yet.
> 
> My main comments are about using pkg-config when cross-compiling. There are also a couple of whitespace issue (trailing spaces).
> 
> Do you mind squashing all your changes into a single commit, to keep the history a bit tidy? I can also take care of that during merging if you prefer.

He Jef,

I think I addressed all your feedback. 

- I squashed all commits into one.
- Used the `env` in github actions instead of those nasty oneliners (which I initially used because it made it easier to debug it locally)
- The libraries are now build and installled in the artifact folder. So it does not use the toolchain directory to put the resulting files and it will  include all required dependencies in the artifact, which is, ama-zing. 
- I also removed some double spaces and trailing spaces

If you have further feedback or I missed some items, let me know!

blackshadev commented on .github/workflows/build.yml:
```diff
@@ -81,8 +81,42 @@ jobs:
+          pushd hidapi-hidapi-${HIDAPI_VERSION}
+          autoreconf --install --force
+          ./configure --host=${{ matrix.arch }}-w64-mingw32  --prefix=/usr
+          make || true
+          make install DESTDIR=$PWD/../artifacts
```
I have added an explanation about that `make || true` to the commit message:

> Errors during making of hidapi are ignored because it will error about determining the host path which is only an issue during uninstallation

Basically it gives an error , but the error is only relevant for "uninstallation" on windows platform.

blackshadev replied:
Error message given by the make of hidapi

```
libtool:   error: Could not determine the host path corresponding to
libtool:   error:   '/home/runner/work/libdivecomputer/libdivecomputer/hidapi-hidapi-0.10.1/windows/.libs'
libtool:   error: Continuing, but uninstalled executables may not work.
libtool:   error: Could not determine the host path corresponding to
libtool:   error:   '/home/runner/work/libdivecomputer/libdivecomputer/hidapi-hidapi-0.10.1/windows/.libs'
libtool:   error: Continuing, but uninstalled executables may not work.
```

jefdriesen replied:
Libtool says "error", but it's not a fatal error that causes the build to fail (by returning a non-zero exit code). So it's more like a warning message. If you look at the libdivecomputer build log, you'll find exactly the same message. It's harmless and can be ignored. The || true construct makes no difference here, and will only ignore any real errors. That's not what we want in a CI setup.

Note that the "uninstalled executables" has nothing to do with uninstallation. It's about running executables directly from the build tree, without installing them first with `make install`. We're not doing that, so the warning is irrelevant.

blackshadev replied:
Ahh check, I didn't know that. I read "error"  as thus failing, so I thought it was necessary. I removed the `|| true` .

blackshadev commented on .github/workflows/build.yml:
```diff
@@ -81,8 +81,42 @@ jobs:
+      - name: Install hidapi
+        env:
+          HIDAPI_VERSION: 0.10.1
+          PKG_CONFIG_LIBDIR: ${{ github.workspace }}/artifacts/usr/lib/pkgconfig
+          PKG_CONFIG_SYSROOT_DIR: ${{ github.workspace }}/artifacts
+          PKG_CONFIG_ALLOW_SYSTEM_CFLAGS: 1
+          PKG_CONFIG_ALLOW_SYSTEM_LIBS: 1
```
I have not tried it without it. But it should depend in libusb. I followed the steps defines in the "[building-hidapi](https://github.com/libusb/hidapi#building-hidapi)" section.i

blackshadev replied:
Lets try if it works without it: https://github.com/blackshadev/libdivecomputer/runs/2739160034

blackshadev replied:
That worked just fine. I removed it.

blackshadev commented on .github/workflows/build.yml:
```diff
@@ -81,8 +81,42 @@ jobs:
+          popd
       - run: autoreconf --install --force
-      - run: ./configure --host=${{ matrix.arch }}-w64-mingw32 --prefix=/usr
+      - name: configure 
```
fixed

jefdriesen commented on .github/workflows/build.yml:
```diff
@@ -81,8 +81,38 @@ jobs:
       - name: Install dependencies
         run: sudo apt-get install gcc-mingw-w64 binutils-mingw-w64 mingw-w64-tools
+      - name: Install libusb
+        env: 
```
There is a trailing space here.

In case you don't know, you can easily visualize whitespace issues with `git show` or `git diff`. They will be marked in red because the default setting for `core.whitespace` is `blank-at-eol,blank-at-eof,space-before-tab`. Another useful one is `git config apply.whitespace fix` to automatically fix them during some git operations.

https://www.git-scm.com/book/en/v2/Customizing-Git-Git-Configuration#_core_whitespace

blackshadev replied:
Sorry for this. Thanks for the suggestions ,with `git show` I saw them clearly. Before this I tried finding them with my editor, which wasn't ideal. I didn't know git showed them like that. Thanks!

jefdriesen commented on .github/workflows/build.yml:
```diff
@@ -81,8 +81,38 @@ jobs:
+          make install DESTDIR=$PWD/../artifacts
+          popd
       - run: autoreconf --install --force
-      - run: ./configure --host=${{ matrix.arch }}-w64-mingw32 --prefix=/usr
```
Same here

jefdriesen commented on .github/workflows/build.yml:
```diff
@@ -81,8 +81,38 @@ jobs:
+          PKG_CONFIG_SYSROOT_DIR: ${{ github.workspace }}/artifacts
+          PKG_CONFIG_ALLOW_SYSTEM_CFLAGS: 1
+          PKG_CONFIG_ALLOW_SYSTEM_LIBS: 1
+        
```
And here. Just remove this line completely.

jefdriesen reviewed (changes requested):
Looks good, just some remaining trailing white space issues in a few places. While you fix those, also remove the comment regarding the hidapi errors from the commit message. It's only confusing.. You are not ignoring any errors because there are no errors.
